src/Magento/FunctionalTestingFramework/Util/ComposerModuleResolver.php

```diff
@@ -0,0 +1,130 @@
+
+    /**
+     * Code path array from composer installed test packages
+     *
+     * @var array
+     */
+    private $installedTestModules = null;
+
+    /**
+     * ComposerModuleResolver constructor
+     */
+    public function __construct()
+    {
+    }
+
+    /**
+     * Get code paths for installed test modules
+     *
+     * @param string $rootComposerFile
+     * @return array
+     * @throws TestFrameworkException
+     */
+    public function getComposerInstalledTestModules($rootComposerFile)
+    {
+        if (null !== $this->installedTestModules) {
+            return $this->installedTestModules;
+        }
+
+        if (!is_file($rootComposerFile) || substr($rootComposerFile, -13) != 'composer.json') {
+            throw new TestFrameworkException("Invalid root composer json file: {$rootComposerFile}");
+        }
+
+        $this->installedTestModules = [];
+        $composerInstaller = new ComposerInstaller(
+            new ComposerFactory($rootComposerFile)
+        );
+
+        foreach ($composerInstaller->getInstalledTestPackages() as $packageName => $packageData) {
+            $suggestedModuleNames = $packageData[ComposerInstaller::KEY_PACKAGE_SUGGESTED_MAGENTO_MODULES];
+            $path = $packageData[ComposerInstaller::KEY_PACKAGE_INSTALLEDPATH];
+            $this->installedTestModules[$path] = $suggestedModuleNames;
+        }
+        return $this->installedTestModules;
+    }
+
+    /**
+     * Get code paths by searching test module composer json file from input directories
+     *
+     * @param array $directories
+     * @return array
+     * @throws TestFrameworkException
+     */
+    public function getTestModulesFromPaths($directories)
+    {
+        if (null !== $this->searchedTestModules) {
+            return $this->searchedTestModules;
+        }
+
+        $this->searchedTestModules = [];
+        foreach ($directories as $directory) {
+            $this->searchedTestModules = array_merge_recursive(
+                $this->searchedTestModules,
+                $this->getTestModules($directory)
+            );
+        }
+        return $this->searchedTestModules;
+    }
+
+    /**
+     * Get code paths by searching test module composer json file from input directory
+     *
+     * @param string $directory
+     * @return array
+     * @throws TestFrameworkException
+     */
+    private function getTestModules($directory)
+    {
+        $normalizedDir = realpath($directory);
+        if (!is_dir($normalizedDir)) {
+            throw new TestFrameworkException("Invalid directory: {$directory}");
+        }
+
+        // Find all composer json files under directory
+        $modules = [];
+        $jsonFinder = new ComposerJsonFinder();
+        $fileList = $jsonFinder->finComposerJsonFilesAtDepth($normalizedDir, 2);
+        foreach ($fileList as $file) {
+            // Parse composer json for test module name and path information
+            $composerInfo = new ComposerPackager(
```

Comment:
Additionally if you are able to remove `ComposerPackager`, I'd get rid of the `AbstractComposerHandler` class as well. As far as I can tell it'd be a collection of getters and a single util function (that can be defined elsewhere).

Reply:
The ComposerPackager class is an object to represent composer.json file. We will work with it a lot in the future, like module dependencies, packaging etc. We might need to add new functions to it. It makes sense to me to be a class itself.

Another consideration is ModuleResolver class is already 880+ lines of code. We actually should consider refactor it to make it more object oriented and shorter but not adding more utility functions into it.

After review the ComposerPackager and AbstractComposerHandler, I have moved parseSuggestsForMagentoModuleNames() in to ComposerPackager where it more belongs.

AbstractComposerHandler is to hold common composer factory property for both classes.

Thoughts?

Reply:
Hm. I see the use in having a `ComposerPackage` (without `r` at end) but I hesitate to add classes until we have a clearly defined need for it, especially since the PR already introduces so many new classes.

I agree with keeping it out of ModuleResolver though, I was thinking you could add the util function to `ComposerModuleResolver` which is still at ~120 lines (if we do end up removing the ComposerPackager class).

I would ask you to just try the following and see the result:
* Remove `ComposerPackager` and `AbstractComposerHandler` and port any required logic to `ComposerModuleResolver` + `ComposerInstaller`. 
* Rename `ComposerInstaller` to `ComposerHandler`

My main goal in any of the suggestions I am making in the PR is to reduce complexity to an absolute minimum; as it is, this PR introduces 6 entirely new classes and completely refactors 2 existing classes (one of which is absolutely central to the framework). I've no doubt that the code works, I just want to make sure we have achieve it in its most straightforward form 👍

Reply:
I forgot to mention that ComposerPackage(r) is for dealing with composer.json which could be installed or not installed and ComposerInstaller is for composer.lock which are installed modules. They also share some seemingly same functions like, to check if a module is a test module, get required and get suggests, check dependencies, etc. But their internal implementation will be different. The former will be parsing the json file, the later will call composer apis. 

Is it a good idea to mingle them into one class?  i.e.
Potential two versions of isMftfTestPackage() in one class?

I know the change is big, but the most complicated part is still in ModuleResolver itself. These 2 classes or 4 classes are minor and I think separate class to deal with composer install package vs composer.json parsed package are more cleaner.

Reply:
It's only a good idea to mingle into one class if we could completely remove one of them, which to your credit is seeming more and more difficult as I dive into the code.

Let me hack away at this and try to see if I can even remove them, I'll need your help for use cases to double check everything still works.